File: uglyNumber2.cpp
/*
//Time: O(n)
//Space:O(n) Space consumed by the priority queue and hashSet
class myCompare{
public:
    bool operator()(long a ,long b){
        return a > b;
    }
};
class Solution {
public:
    int nthUglyNumber(int n) {
        set<long> hSet;
        priority_queue<long,vector<long>,myCompare> pq; //min-heap
        hSet.insert(1);
        pq.push(1);
        vector<int> primes = {2,3,5};
        int i = 0;
        long curr =0;
        while(i < n){
            curr= pq.top();
            pq.pop();
            for(int i = 0;i < 3;i++){
                long num = curr*primes[i];
                if(!hSet.count(num)){
                    hSet.insert(num);
                    pq.push(num);
                }
            }
            i++;
        }
        return (int)curr;
    }
};
*/
//Time:O(n)
//Space:O(1)
class Solution {
public:
    int nthUglyNumber(int n) {
        vector<int> ans(n);
        ans[0] = 1;
        int n2 = 2,n3 = 3,n5 = 5;
        int p5 = 0,p2 = 0,p3 =0;
        for(int i = 1; i < n; i++){
            int minVal = min(n2,min(n3,n5));
            ans[i] = minVal;
            if(n2 == minVal){
                p2++;
                n2= 2 * ans[p2];
            }
            if(n3 == minVal){
                p3++;
                n3=3 * ans[p3];
            }
            if(n5 == minVal){
                p5++;
                n5=5 * ans[p5];
            }
        }
        return ans[n-1];
    }
};

File: ugly_number_2.py
# Time Complexity : O(n)
# Space Complexity: O(n)

class Solution:
    def nthUglyNumber(self, n: int) -> int:
        if n == 1:
            return 1
        res = [0 for _ in range(n)]
        res[0] = 1
        p2 = 0
        p3 = 0
        p5 = 0
        n2 = 2
        n3 = 3
        n5 = 5
        for i in range(1, n):
            # print(res)
            min_val = min(n2, min(n3, n5))
            res[i] = min_val
            if min_val == n2:
                p2 += 1
                n2 = res[p2]*2
            if min_val == n3:
                p3 += 1
                n3 = res[p3]*3
            if min_val == n5:
                p5 += 1
                n5 = res[p5]*5

        return res[n-1]

#         visited = set()
#         nums = []
#         heapq.heapify(nums)
#         heapq.heappush(nums, 1)
#         visited.add(1)
#         out = []
#         count = 0
#         while count < n:
#             val = heapq.heappop(nums)
#             out.append(val)
#             count += 1
#             for i in [2,3,5]:
#                 new = val * i
#                 if new not in visited:
#                     heapq.heappush(nums, new)
#                     visited.add(new)
#         print(out)
#         return out[-1]
